daemon/daemon_utils.go

```diff
@@ -52,3 +59,93 @@ func initialRuntime(baseDir string, runtimes map[string]types.Runtime) error {
 
 	return nil
 }
+
+// checkSnapshotter checks whether the given snapshotter is valid
+func checkSnapshotter(snapshotter string, containerdRootDir string) error {
+	var (
+		driverFound = false
+	)
+
+	fInfo, err := os.Stat(containerdRootDir)
+	if err != nil {
+		return fmt.Errorf("failed to stat containerd root dir: %v", err)
+	}
+
+	if !fInfo.IsDir() {
+		return fmt.Errorf("containerd root dir %s is not a directory", containerdRootDir)
+	}
+
+	infos, err := ioutil.ReadDir(containerdRootDir)
+	if err != nil {
+		return fmt.Errorf("failed to read dir %s: %v", containerdRootDir, err)
+	}
+
+	// read all snapshotter dir to make sure snapshotter is valid
+	for _, info := range infos {
+		found, err := checkSnapshotterDir(snapshotter, containerdRootDir, info)
+		if err != nil {
+			return err
+		}
+
+		if found {
+			driverFound = found
+		}
+	}
+
+	if !driverFound {
+		return fmt.Errorf("containerd not support snapshotter driver %s", snapshotter)
+	}
+
+	return nil
+}
+
+func checkSnapshotterDir(snapshotter string, containerdRootDir string, info os.FileInfo) (driverFound bool, err error) {
+	if !info.IsDir() {
+		return
+	}
+
+	name := info.Name()
+
+	if !strings.HasPrefix(name, storageDirPrefix) {
+		return
+	}
+
+	driver := name[len(storageDirPrefix):]
+
+	if snapshotter == driver {
+		driverFound = true
+		return
+	}
+
+	// check other snapshotter snapshots dirs to make sure current snapshotter driver is not other
+	snapshotsDir := filepath.Join(containerdRootDir, name, "snapshots")
+	if _, er := os.Lstat(snapshotsDir); er != nil {
+		if os.IsNotExist(er) {
+			return
+		}
+
+		logrus.Warnf("failed to stat snapshots dir of driver %s: %v", driver, er)
+		return
+	}
+
+	snapshots, er := ioutil.ReadDir(snapshotsDir)
+	if er != nil {
+		logrus.Warnf("failed to read snapshots of driver %s: %v", driver, er)
+		return
+	}
+
+	for _, s := range snapshots {
+		if !s.IsDir() {
+			return
+		}
+
+		// if dir name is the number, we consider this driver as current snapshotter driver
+		_, er = strconv.Atoi(s.Name())
+		if er == nil {
+			err = fmt.Errorf("current snapshotter driver is %s, cannot change to snapshotter driver %s", driver, snapshotter)
+			return
```

Comment:
this function just return error at here,  other return err is nil, this logic looks a little strange, can you give some suggestions? @fuweid

Reply:
The style different with others. I do not see `driverFound bool, err error) ` make code simpler, @wangforthinker

Reply:
Now, it is outdated!